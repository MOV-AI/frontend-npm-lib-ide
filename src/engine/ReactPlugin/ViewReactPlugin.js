import React from "react";
import IDEPlugin from "../IDEPlugin/IDEPlugin";
import PluginManagerIDE from "../PluginManagerIDE/PluginManagerIDE";
import { getRefComponent } from "../../utils/Utils";

export class ViewReactPlugin extends IDEPlugin {
  /**
   *
   * @returns {React Component}
   */
  render() {
    // Abstract method should be implemented in subclasses
  }

  async activate() {
    if (this.profile.location)
      await this.call(
        this.profile.location,
        "addView",
        this.profile,
        this.render(),
      );
    super.activate();
  }

  deactivate() {
    if (this.profile.location)
      this.call(this.profile.location, "removeView", this.profile);
    super.deactivate();
  }
}

export class ViewPlugin extends ViewReactPlugin {
  constructor(profile, props = {}, methods = []) {
    const existingPlugin = PluginManagerIDE.getPlugin(profile.name);
    if (existingPlugin) return existingPlugin;
    super({
      ...profile,
      methods: ["render", ...methods],
    });

    this.methods = methods;
    this.ref = React.createRef();
    this.initMethods();
    this.props = props;
  }

  initMethods = () => {
    this.methods.forEach((name) => {
      this[name] = (...a) => {
        if (!this.ref.current) {
          return console.warn("debug method not implemented in component");
        }
        return this.ref.current[name](...a);
      };
    });
  };
}

/**
 *
 * @param {ReactComponent} ReactComponent
 * @returns {ViewReactPlugin}
 */
export function withViewPlugin(ReactComponent, methods = []) {
  const RefComponent = getRefComponent(ReactComponent);
  return class extends ViewPlugin {
    constructor(profile, props = {}) {
      super(profile, props, methods);
    }

    render(otherProps = {}) {
      return (
        <RefComponent
          {...this.props}
          {...otherProps}
          ref={this.ref}
          call={this.call}
          profile={this.profile}
          emit={this.emit}
          off={this.off}
          on={this.on}
          onTopic={this.onTopic}
        />
      );
    }
  };
}

//========================================================================================
/*                                                                                      *
 *                                   Shared hooks                                      *
 *                                                                                      */
//========================================================================================

/**
 * Hook to allow use of methods in view plugins
 * @param {ReactRef} ref : React ref
 * @param {Object} methods : Object maping all the methods to be exported in the plugin
 */
export const usePluginMethods = (ref, methods) => {
  React.useImperativeHandle(
    ref,
    () => ({
      ...methods,
    }),
    [methods],
  );
};

/**
 * Give the previous state of a prop or state hook on change
 * @param {*} value : Value to get previous state
 * @returns {*} Previous value
 */
export const usePrevious = (value) => {
  const ref = React.useRef();
  React.useEffect(() => {
    ref.current = value;
  });
  return ref.current;
};
