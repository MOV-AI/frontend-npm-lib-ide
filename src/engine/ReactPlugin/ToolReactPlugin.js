import React, { forwardRef, useEffect } from "react";
import { withTheme } from "@mov-ai/mov-fe-lib-react";
import { makeStyles } from "@mov-ai/mov-fe-lib-react";
import { drawerSub } from "../../plugins/hosts/DrawerPanel/DrawerPanel";
import withAlerts from "../../decorators/withAlerts";
import withMenuHandler from "../../decorators/withMenuHandler";
import { PLUGINS } from "../../utils/Constants";
import { composeDecorators } from "../../utils/Utils";
import { ViewPlugin } from "./ViewReactPlugin";

export const useStyles = makeStyles(() => ({
  root: {
    height: "100%"
  }
}));

/**
 * Decorate react component and handle shared behavior between editors
 * @param {ReactComponent} ReactComponent : Editor React Component
 * @param {Array<String>} methods : Methods to be exposed
 * @returns
 */
export function withToolPlugin(ReactComponent, methods = []) {
  const RefComponent = forwardRef((props, ref) => ReactComponent(props, ref));

  const ToolComponent = forwardRef((props, ref) => {
    const { on, off, profile } = props;

    /**
     * Component did mount
     */
    useEffect(() => {
      drawerSub.url = profile.name;
      on(
        PLUGINS.TABS.NAME,
        PLUGINS.TABS.ON.ACTIVE_TAB_CHANGE,
        async ({ id }) => {
          if (profile.name === id)
            drawerSub.url = profile.name;
        }
      );

      return () => {
        off(PLUGINS.TABS.NAME, PLUGINS.TABS.ON.ACTIVE_TAB_CHANGE);
      };
    }, [profile.name]);

    return <RefComponent {...props} ref={ref} />;
  });

  // Decorate component
  const DecoratedToolComponent = composeDecorators(ToolComponent, [
    withTheme,
    withAlerts,
    withMenuHandler
  ]);

  /**
   * Return Plugin class rendering decorated editor component
   */
  return class extends ViewPlugin {
    constructor(profile, props = {}) {
      super(profile, props, methods);
    }

    render() {
      return (
        <DecoratedToolComponent
          {...this.props}
          ref={this.ref}
          call={this.call}
          profile={this.profile}
          emit={this.emit}
          on={this.on}
          off={this.off}
          onTopic={this.onTopic}
        />
      );
    }
  };
}
