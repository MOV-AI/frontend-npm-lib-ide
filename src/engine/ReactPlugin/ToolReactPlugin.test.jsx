/* eslint-env jest */

import React from "react";
import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { withToolPlugin } from "./ToolReactPlugin";
import { ViewPlugin } from "./ViewReactPlugin";

// Mock decorators
jest.mock("@mov-ai/mov-fe-lib-react", () => ({
  withTheme: (Component) => Component,
}));
jest.mock("../../decorators/withAlerts", () => (Component) => Component);
jest.mock("../../utils/Utils", () => ({
  composeDecorators: (Comp, decorators) =>
    decorators.reduce((C, dec) => dec(C), Comp),
}));

// Mock keybinds util
jest.mock("../../utils/keybinds", () => ({
  setUrl: jest.fn(),
}));

// Mock Constants
jest.mock("../../utils/Constants", () => ({
  PLUGINS: {
    TABS: {
      NAME: "tabs",
      CALL: { GET_ACTIVE_TAB: "getActiveTab", FOCUS_EXISTING_TAB: "focusTab" },
    },
    ORCHESTRATOR: {
      NAME: "orchestrator",
      CALL: { RENDER_MENUS: "renderMenus" },
    },
  },
}));

describe("withToolPlugin HOC", () => {
  const DummyComponent = jest.fn(() => <div data-testid="dummy">Hello</div>);

  it("returns a ViewPlugin subclass whose render outputs the inner component", () => {
    const PluginClass = withToolPlugin(DummyComponent);
    expect(PluginClass.prototype).toBeInstanceOf(ViewPlugin);
    const instance = new PluginClass({ name: "testPlugin" }, {});
    const fakeCall = jest.fn();

    // Wrap in a role=tabpanel container, override call prop
    const { getByTestId } = render(
      <div role="tabpanel" id="tab-1">
        {React.cloneElement(instance.render(), { call: fakeCall })}
      </div>,
    );

    expect(getByTestId("dummy")).toHaveTextContent("Hello");
  });

  it("calls renderMenus, focusExistingTab and activateTool on mount", async () => {
    const fakeCall = jest.fn().mockResolvedValueOnce({ id: "tab-1" });
    const profile = { name: "profile1" };
    const PluginClass = withToolPlugin(DummyComponent);
    const instance = new PluginClass(profile, {});
    const element = instance.render();

    render(
      <div role="tabpanel" id="tab-1">
        {React.cloneElement(element, { call: fakeCall })}
      </div>,
    );

    await waitFor(() => {
      expect(fakeCall).toHaveBeenCalledWith(
        "orchestrator",
        "renderMenus",
        expect.objectContaining({ id: "tab-1", ref: instance.ref }),
      );
      expect(fakeCall).toHaveBeenCalledWith("tabs", "focusTab", "tab-1");
    });
  });

  it("re-activates tool on click", async () => {
    const fakeCall = jest.fn().mockResolvedValue({ id: "tab-2" });
    const profile = { name: "profile2" };
    const PluginClass = withToolPlugin(DummyComponent);
    const instance = new PluginClass(profile, {});
    const element = instance.render();

    const { getByTestId } = render(
      <div role="tabpanel" id="tab-2">
        {React.cloneElement(element, { call: fakeCall })}
      </div>,
    );

    userEvent.click(getByTestId("dummy").parentElement);

    await waitFor(() => {
      expect(fakeCall).toHaveBeenCalledWith("tabs", "getActiveTab");
    });
  });

  it("matches snapshot via RTL", () => {
    const PluginClass = withToolPlugin(() => <span>Snap</span>);
    const instance = new PluginClass({ name: "snap" }, {});
    const fakeCall = jest.fn();

    const { asFragment } = render(
      <div role="tabpanel" id="snap-tab">
        {React.cloneElement(instance.render(), { call: fakeCall })}
      </div>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
