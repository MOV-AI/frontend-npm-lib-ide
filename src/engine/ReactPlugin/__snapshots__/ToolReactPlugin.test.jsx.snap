// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withToolPlugin HOC matches snapshot via RTL 1`] = `
<DocumentFragment>
  <div
    id="snap-tab"
    role="tabpanel"
  >
    <div
      style="height: 100%;"
      tabindex="-1"
    >
      <span>
        Snap
      </span>
    </div>
  </div>
</DocumentFragment>
`;
